### ISSUE TITLE
installation without homebrew on MacOS

### ISSUE THREAD

**Issue by jar398** [gnames/gnparser#274]
This is a continuation of #236.  Once again I started from the README, section https://github.com/gnames/gnparser?tab=readme-ov-file#linux-or-mac-os-x, and the method proposed there did not work (and I can't figure out how to make it work since MacOS has removed the override mechanism that I used before).  May I suggest a qualifier in that section saying something to the effect that ever-changing MacOS security measures may make this installation method difficult or impossible.  Or if you can't find a way to instruct on how to install successfully in modern MacOS (I'm running Sonoma 14.5) perhaps you should just remove this method in favor of homebrew (which I am about to try).

**Comment by dimus** [comment]
thanks for letting us know @jar398. Can it be that /usr/local/bin is not in your PATH? I tried the following on mac:

```bash
~/Downloads
09:26:40 ❯ wget https://github.com/gnames/gnparser/releases/download/v1.11.0/gnparser-v1.11.0-mac-arm.tar.gz
--2024-11-12 09:27:45--  https://github.com/gnames/gnparser/releases/download/v1.11.0/gnparser-v1.11.0-mac-arm.tar.gz
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/320967495/b8a752c9-c9a2-4be8-a2d8-d9ab9646f32b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241112%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241112T092745Z&X-Amz-Expires=300&X-Amz-Signature=433e4aaa6990489c5d04582b4434bb63b5fe145be9acb1807998ee3bd6036606&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dgnparser-v1.11.0-mac-arm.tar.gz&response-content-type=application%2Foctet-stream [following]
--2024-11-12 09:27:45--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/320967495/b8a752c9-c9a2-4be8-a2d8-d9ab9646f32b?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20241112%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241112T092745Z&X-Amz-Expires=300&X-Amz-Signature=433e4aaa6990489c5d04582b4434bb63b5fe145be9acb1807998ee3bd6036606&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3Dgnparser-v1.11.0-mac-arm.tar.gz&response-content-type=application%2Foctet-stream
Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4630373 (4.4M) [application/octet-stream]
Saving to: ‘gnparser-v1.11.0-mac-arm.tar.gz’

gnparser-v1.11.0-mac-arm.tar. 100%[================================================>]   4.42M  20.2MB/s    in 0.2s

2024-11-12 09:27:45 (20.2 MB/s) - ‘gnparser-v1.11.0-mac-arm.tar.gz’ saved [4630373/4630373]


~/Downloads
09:27:45 ❯ tar xvf gnparser-v1.11.0-mac-arm.tar.gz
gnparser

~/Downloads
09:28:12 ❯ sudo cp gnparser /usr/local/bin

~/Downloads
09:28:17 ❯ which gnparser
/usr/local/bin/gnparser

~/Downloads
09:28:24 ❯ gnparser "Bubo bubo"
Id,Verbatim,Cardinality,CanonicalStem,CanonicalSimple,CanonicalFull,Authorship,Year,Quality
4431a0f3-e901-519a-886f-9b97e0c99d8e,Bubo bubo,2,Bubo bub,Bubo bubo,Bubo bubo,,,1
```

**Comment by dimus** [comment]
I did not update gnparser to newest verwion on homebrew yet, I  want to fix some small things and release v1.11.1 today and then update homebrew.

**Comment by dimus** [comment]
Is your mac x86_64 or silicon @jar398?

**Comment by dimus** [comment]
out of curiosity I did try brew method on 2013 mad as well. It took quilte a long time, but it did finish:

```bash
~/Downloads took 26s
09:44:33 ❯ brew tap gnames/gn
==> Auto-updating Homebrew...
Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with
HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Downloading https://ghcr.io/v2/homebrew/portable-ruby/portable-ruby/blobs/sha256:720d4fb1164e600f787d656019a8e46314dc38e1885f4a8df809c180acf5e7b3
################################################################################################################ 100.0%
==> Pouring portable-ruby-3.3.6.el_capitan.bottle.tar.gz
==> Auto-updated Homebrew!
Updated 5 taps (homebrew/bundle, homebrew/services, gnames/gn, homebrew/core and homebrew/cask).
==> New Formulae
action-docs         decasify            kanata              markdown-oxide      rip2                vipsdisp
azqr                dipc                langgraph-cli       mkdocs-material     tcl-tk@8            wasi-runtimes
boring              icu4c@74            libmsquic           onnx                toml2json           wasm-component-ld
comtrya             icu4c@76            libtatsu            pytest              typos-lsp           zizmor
==> New Casks
acronis-true-image-cleanup-tool         fathom                                  lunatask
aide-app                                font-eldur                              mailsteward
ba-connected                            font-parkinsans                         neohtop
bobhelper                               font-recursive-desktop                  viz
claude                                  github-copilot-for-xcode                whimsical
default-handler                         hyperconnect                            yellowdot
djuced                                  jet-pilot
==> Deleted Installed Formulae
pkg-config ✘

You have 14 outdated formulae installed.


~/Downloads took 14s
09:53:04 ❯ brew install gnparser
gnparser 1.10.3 is already installed but outdated (so it will be upgraded).
Warning: You are using macOS 10.15.
We (and Apple) do not provide support for this old version.
It is expected behaviour that some formulae will fail to build in this old version.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.

==> Fetching dependencies for gnames/gn/gnparser: go
==> Fetching go
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/e27d8ab9decdc2f84b00ecc1124c21a0146162e7/Formu
################################################################################################################ 100.0%
==> Downloading https://storage.googleapis.com/golang/go1.20.14.darwin-amd64.tar.gz
Already downloaded: /Users/dimus/Library/Caches/Homebrew/downloads/4c20a76e01f813497d7f320849ccf512626090bd70bda6e1c655134da639e113--go1.20.14.darwin-amd64.tar.gz
==> Downloading https://go.dev/dl/go1.23.3.src.tar.gz
==> Downloading from https://dl.google.com/go/go1.23.3.src.tar.gz
################################################################################################################ 100.0%
==> Fetching gnames/gn/gnparser
==> Downloading https://github.com/gnames/gnparser/archive/v1.10.4.tar.gz
==> Downloading from https://codeload.github.com/gnames/gnparser/tar.gz/refs/tags/v1.10.4
#-#O=-#    #
==> Upgrading gnames/gn/gnparser
  1.10.3 -> 1.10.4
==> Installing dependencies for gnames/gn/gnparser: go
==> Installing gnames/gn/gnparser dependency: go
==> ./make.bash
==> /usr/local/Cellar/go/1.23.3/bin/go install std cmd
🍺  /usr/local/Cellar/go/1.23.3: 13,235 files, 272.2MB, built in 3 minutes 39 seconds
==> Installing gnames/gn/gnparser
==> go build -o ./gnparser
🍺  /usr/local/Cellar/gnparser/1.10.4: 8 files, 17.4MB, built in 9 seconds
==> Running `brew cleanup gnparser`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Removing: /usr/local/Cellar/gnparser/1.10.3... (8 files, 12.4MB)
Removing: /Users/dimus/Library/Caches/Homebrew/gnparser--1.10.3.tar.gz... (468.9KB)
```

**Comment by dimus** [comment]
@mjy points that on mac the following settings need to be enabled:
<img width="719" alt="image" src="https://github.com/user-attachments/assets/27691444-ecd1-4d6a-a1ed-7715319c2c44">

**Comment by jar398** [comment]
My computer is an Air with an m2 (ARM) processor and MacOS 14.5.  I was not having a PATH problem, since the error was a security issue not a file not found issue.  The homebrew method works just fine so that is what I'm using now.

The Developer Tools method looks great, thanks.  If you endorse this for gnparser I suggest you change the instructions to say XCode must be installed and a security override should be enabled before attempting to run non-homebrew gnparser.  If you don't want to do this, I recommend removing the instructions for non-homebrew installation since they won't work as currently written.

**Comment by dimus** [comment]
Thanks for the info @jar398 . Can you confirm, that adding the developer tools' permission allowed you to execute binary gnparser file directly, if yes, I will add it to the documentation.

Xcode is not needed for execution of the binary files (compiled Go, C, C++, Rust programs), but is required for brew as a tool to work correctly, because brew compiles its packages by itself. I think installing and configuring homebrew is beyond the scope of gnparser documentation, as it is already provided by homebrew web-site and in help when running brew commands.

**Comment by dimus** [comment]
I added:

```
Please note that Mac OS might prevent running executables that are not
'signed' up officially by Apple. To overcome this problem go to
`System Settings`, search for `developer tools`and under`Allow applications
below to run software locally that does not meet the system's security policy`
add the terminal emulation software that you normally use (e.g. Terminal app).
```

**Comment by jar398** [comment]
The instructions you gave for allowing execution of gnparser did not work for me.  I enabled Terminal in Settings / Privacy &Security / Allow Applications (etc.), opened a new Terminal window, and tried to run the ARM gnparser from the release file.  I got the same error as before ('Apple cannot check it for malicious software').

I could read more about security policies but do not have that kind of patience now.  (I wonder what homebrew does in order to bypass this check.)

**Comment by dimus** [comment]
hm,

**Comment by dimus** [comment]
I think I got the error finally. I created a new account, 

1. downloaded tar file, extracted gnparser.

2. When I tried to run it in a terminal I got a dialog box with `"gnparser" cannot be opened ....` with two buttons `Done` and `Move gnparser to trash`.

3. I click "Done"

4. I go to System Settings -> Privacy & Security and scroll to the bottom

5.  In the Security sextion after "Allow applications from"... I see `"gnparser" was blocked to protect your Mac` with button "Allow anyway"

6. I click Allow anyway and go back to terminal

7. I try to execute gnparser again from teminal, Now I see dialog with 3 buttons: "Move to Trash", "Open anyway", "Done"

8. I click "Open anyway", it asks for admin username and password.

9. I enter them and after that the executable is unblocked

10. After that gnparser binary can be moved to another directory, renamed, it still works as executable.

@jar398 can you try to follow similar steps and see if it works for you? If yes, then I will add the key point to documentation -- allowing in System Settings to open gnparser.

**Comment by jar398** [comment]
sorry for the delay, I'll get to this.

**Comment by dimus** [comment]
@jar398 did you have a chance to try reproduce the steps? Can we close this issue?

**Comment by dimus** [comment]
To my knowledge the documentation is now up to date. @jar398 feel free to reopen it, if the problem reappear.

**Comment by jar398** [comment]
Sorry I dropped this on the floor.  I did get it running somehow, as did a friend who had an old mac, but the friend had to read the conversation above in order to get to where he could run gnparser without homebrew (he uses Macports; not sure what he did but he did not need my or your help).  Since things are working on my Mac, and tracking down the particular failure so that you can hone your documentation for old OSes will be time consuming, I think I will leave this alone.

Thanks for your help.

**Comment by dimus** [comment]
Sadly MacOS is getting crazier with every new update in its paranoia and mistrust that people know what they are doing. So far homebrew was the most reliable way to get gn things on mac. The good thing about old hardware -- Apple does not support it anymore, and old macos is much less annoying security-wise. I am glad gnparser install did work for you, so yes, Ill keep docs the way they are, and update them when they will become outdated again.
